Add micro benchmarks for serialising and applying Transactions

abailly-iohk opened the pull request:
Following discussions on #186 this PR adds a few micro-benchmarks to clarify what is the performance bottleneck for applying transactions to Hydra Head. Here are the results on a recent MBP with M2 chip:

```
Running 1 benchmarks...
Benchmark micro: RUNNING...
benchmarking Cardano Ledger/Apply Tx
time                 281.6 μs   (280.6 μs .. 282.7 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 281.3 μs   (280.9 μs .. 281.9 μs)
std dev              1.536 μs   (969.4 ns .. 2.170 μs)

benchmarking Cardano Ledger/Serialize NewTx (JSON)
time                 63.98 μs   (63.63 μs .. 64.61 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 63.83 μs   (63.69 μs .. 64.44 μs)
std dev              743.5 ns   (165.1 ns .. 1.671 μs)

benchmarking Cardano Ledger/Serialize NewTx (CBOR)
time                 1.164 μs   (1.163 μs .. 1.166 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.164 μs   (1.163 μs .. 1.166 μs)
std dev              3.917 ns   (2.369 ns .. 5.681 ns)

benchmarking Cardano Ledger/Deserialize NewTx (JSON)
time                 115.0 μs   (114.7 μs .. 115.1 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 114.6 μs   (114.5 μs .. 114.8 μs)
std dev              391.6 ns   (286.9 ns .. 582.4 ns)

benchmarking Cardano Ledger/Deserialize NewTx (CBOR-in-JSON)
time                 11.72 μs   (11.70 μs .. 11.74 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 11.70 μs   (11.69 μs .. 11.71 μs)
std dev              36.75 ns   (27.25 ns .. 57.05 ns)

Benchmark micro: FINISH
```

Interestingly, deserialising a JSON transaction takes about half the time of applying it so it seems we have some progression margin to improve the overall head performance.


---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [?] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

abailly committed:
Add micro-benchmark assessing Tx application time

abailly committed:
Add serialisation time to micro-benchmark

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -239,6 +238,27 @@ benchmark tx-cost
 
+benchmark micro
+  import:         project-config
+  hs-source-dirs: exe/micro-bench
```
We usually keep benchmarks in the `bench/` directory within a package.

Should: use `bench/` as a source directory

ch1bo replied:
You're right. `tx-cost` is different than the other 3 benchmarks we have on the repository. I'd say we should move all to `bench/`

ch1bo commented on hydra-node/exe/micro-bench/Main.hs:
```diff
@@ -0,0 +1,89 @@
+      }
+
+defaultGlobals :: Ledger.Globals
+defaultGlobals =
```
Should: not create another location for fixtures

This and other symbols in here seems to have been copied from the hydra-node tests component. I'm not sure what weighs more, but maybe we want to have only a single set of fixtures? The `epochInfo` and `pparams` fixtures are coming from the `hydra-node` library already. Maybe we can create a more clear fixtures module which would make things less redundant and still not accidentally used?

ch1bo reviewed (changes requested):
Having these benchmarks will become handy.

Besides the comments below, should we run them in CI?

ghost commented:
re-publication of benchmarks: It's trivial to output an html page so we could definitely do it, with the usual caveat those benchmarks are only meaningful to compare things as absolute values produced from running on a random VM environment are probably meaninglees.

abailly committed:
Move benchmarks to bench/ directory

ch1bo commented:
> re-publication of benchmarks: It's trivial to output an html page so we could definitely do it, with the usual caveat those benchmarks are only meaningful to compare things as absolute values produced from running on a random VM environment are probably meaninglees.

I don't mean to publish them, maybe just run them in CI first. But that would increase duration of it. That was just a thought, I'm fine with not doing anything about it.

abailly committed:
Move Globals and default ledger env to shared Fixture module

abailly committed:
Publish micro-benchmarks results

ghost commented:
running them without publishing or _acting_  on them in some way is not very useful. I have added a job to publish them which should not add significant delay to the CI.

abailly committed:
Fix micro benchmark run in Nix

abailly committed:
Fix publication of micro-benchmarks results

abailly committed:
Unneeded package in micro

abailly committed:
Cleanup micro benchmark's build

ch1bo reviewed (approved):
Approved, might need some fixing with recently changed way of assembling docs.

pgrange committed:
Let's put benchmark straight in the target directory

That way, we don't have to move the things later... although
I'm not happy with the bench having to know too much about the
doc structure but...
